urwid/str_util.py

```diff
@@ -204,17 +236,25 @@ def calc_width(text: str | bytes, start_offs: int, end_offs: int) -> int:
     if start_offs > end_offs:
         raise ValueError((start_offs, end_offs))
 
-    utfs = isinstance(text, bytes) and _byte_encoding == "utf8"
-    unis = not isinstance(text, bytes)
-    if (unis and not SAFE_ASCII_RE.match(text)) or (utfs and not SAFE_ASCII_BYTES_RE.match(text)):
-        decode = [decode_one, decode_one_uni][unis]
+    if isinstance(text, str):
+        return sum(get_char_width(char) for char in text[start_offs:end_offs])
+
+    if _byte_encoding == "utf8":
+        try:
+            return sum(get_char_width(char) for char in text[start_offs:end_offs].decode("utf-8"))
+        except UnicodeDecodeError as exc:
```

Comment:
yes, we really sometimes hit this due to splitting in the middle of character.

Reply:
The solution is to use the [codecs.IncrementalDecoder()](https://docs.python.org/3/library/codecs.html#codecs.IncrementalDecoder) class.

Related, about using wcwidth, I wanted to also tell you that because of "grapheme clustering" it is better to use `wcwidth.wcswidth()` rather than `sum( ... wcwidth.wcwidth() )`, because there are character *sequences* which are measured differently than the sum of individual codepoints, for example,

```
>>> wcwidth.wcswidth('\u2665\uFE0F')
2
>>> wcwidth.wcwidth('\u2665') + wcwidth.wcwidth('\uFE0F')
1
```

However, Variation Selector 16 ([U+FE0F](https://codepoints.net/U+FE0F)) has very poor support in terminals, less than half of popular terminals support it, so I wouldn't worry about it too much! I test all popular terminals with their results here, https://ucs-detect.readthedocs.io/results.html

Reply:
>  The solution is to use the [codecs.IncrementalDecoder()](https://docs.python.org/3/library/codecs.html#codecs.IncrementalDecoder) class.

`start_offs` maybe already in the middle of character. I prefer to fix it permanently in the future changes with full switch of internals to the `str` as it's only one supported output for the `RawDisplay`, `CursesDisplay` in case of UTF8 mode and for web/HTML displays.

And about `wcswidth` I added comment, that we need to support compatible "trim to the column" first, or we can have render issues / hard performance issues